File: ios/SharedList/Core/Networking/APIClient.swift
import Foundation

// APIClient — single point of network access for the iOS app.
//
// Responsibilities, in order of importance:
//   1. Send Codable bodies, decode Codable responses, surface typed errors.
//   2. Inject `Authorization: Bearer <accessToken>` from TokenStore.
//   3. On a 401, refresh the access token AT MOST ONCE PER REQUEST and retry,
//      with multiple concurrent in-flight requests sharing one refresh.
//
// Concept #3 is the "single-flight refresh" pattern. Why it matters:
//
//   - Suppose ten requests fire from ten ViewModels at app foreground.
//   - The access token has expired since last use.
//   - All ten get a 401.
//   - Without single-flight, all ten would call /auth/refresh in parallel.
//     The backend's refresh endpoint rotates the refresh token (new one issued,
//     old one marked used). The first request swaps the token, but the other
//     nine race using the same OLD refresh token. The backend's reuse
//     detection then revokes ALL refresh tokens for the user — instant logout.
//   - With single-flight: the first 401 starts a single refresh `Task`; every
//     other concurrent 401 awaits that same Task; when it completes, all
//     callers retry their original request with the new access token.
//
// The implementation here uses an actor (`RefreshCoordinator`) to protect the
// "is a refresh in flight" state. An actor serializes its own methods, so two
// concurrent calls to `refreshIfNeeded()` will see each other; the first
// starts the Task and stores it, the second sees the stored Task and awaits
// the same value. Once the Task completes, the coordinator nils out the
// reference so the next 401 can trigger a fresh refresh.
//
// Note: we deliberately do NOT proactively check token expiry before each
// request. There are two reasons:
//   1. The access token's `exp` is on the server; we'd have to re-derive it
//      from the JWT or cache the issued time, both of which add code without
//      changing the user-visible behavior.
//   2. Reactive (refresh on 401) handles the correct cases (clock skew, server
//      clock drift, sleep/wake jitter) for free.
//
// What we WILL want eventually but not this phase:
//   - Request cancellation propagation (cancelling the calling Task should
//     cancel the URLSessionTask). URLSession cooperates with structured
//     cancellation already; we just need to make sure we don't shield it.
//   - Retry-after handling on 429 (rate-limited paths). Not needed for v1.

public enum APIError: Error, Sendable, Equatable {
    case notAuthenticated
    case server(status: Int, code: String, message: String, requestId: String)
    case decoding(String)
    case transport(message: String)
    case refreshFailed
}

// Protocol so tests can inject a fake. We only model the surface area we
// actually use; URLSession itself conforms to this with no extra glue.
public protocol HTTPRequesting: Sendable {
    func data(for request: URLRequest) async throws -> (Data, URLResponse)
}

extension URLSession: HTTPRequesting {}

public final class APIClient: Sendable {
    public let baseURL: URL
    private let session: any HTTPRequesting
    private let tokenStore: TokenStore
    private let refresher: RefreshCoordinator
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    public init(baseURL: URL, session: any HTTPRequesting = URLSession.shared, tokenStore: TokenStore) {
        self.baseURL = baseURL
        self.session = session
        self.tokenStore = tokenStore
        self.refresher = RefreshCoordinator()
        // The backend serializes timestamps as ISO 8601 strings; we'll need
        // .iso8601 once we start decoding `updated_at`. Phase 5 only sees auth
        // payloads which use plain strings, but configuring it now means
        // future Phase 7 work doesn't change a global decoder.
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        self.decoder = decoder
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        self.encoder = encoder
    }

    // Public entry point. The generic `Body: Encodable` covers both nil-body
    // (GET) and JSON-body (POST/PATCH) requests; passing `EmptyBody()` for GETs
    // keeps the call sites readable. `Response: Decodable` covers responses;
    // `EmptyResponse` for 204 endpoints.
    //
    // `requiresAuth` controls whether we add the bearer token. Login/signup
    // are unauth; everything else is auth. Refresh has a special path inside
    // `RefreshCoordinator` because it must NOT itself trigger another refresh.
    public func send<Body: Encodable & Sendable, Response: Decodable & Sendable>(
        method: String,
        path: String,
        body: Body? = nil,
        requiresAuth: Bool = true,
        responseType: Response.Type = Response.self
    ) async throws -> Response {
        let request = try await buildRequest(method: method, path: path, body: body, requiresAuth: requiresAuth)
        let (data, response) = try await performWithRefresh(request: request, requiresAuth: requiresAuth)
        return try decode(data: data, response: response)
    }

    // Send variant that returns no body (204). Decoding `Void` is awkward in
    // Swift, so we model it as a separate entry point rather than torturing
    // generics. The data is read but ignored.
    public func sendNoContent<Body: Encodable & Sendable>(
        method: String,
        path: String,
        body: Body? = nil,
        requiresAuth: Bool = true
    ) async throws {
        let request = try await buildRequest(method: method, path: path, body: body, requiresAuth: requiresAuth)
        let (data, response) = try await performWithRefresh(request: request, requiresAuth: requiresAuth)
        guard let http = response as? HTTPURLResponse else {
            throw APIError.transport(message: "non-HTTP response")
        }
        if !(200..<300).contains(http.statusCode) {
            throw Self.decodeErrorEnvelope(data: data, status: http.statusCode)
        }
    }

    // MARK: - Internals

    private func buildRequest<Body: Encodable & Sendable>(
        method: String,
        path: String,
        body: Body?,
        requiresAuth: Bool
    ) async throws -> URLRequest {
        // Treat the path as a relative path; URL(string:relativeTo:) handles
        // the join. Paths must start with "/" for predictable behavior.
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw APIError.transport(message: "invalid url for path: \(path)")
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body, !(body is EmptyBody) {
            request.httpBody = try encoder.encode(body)
        }
        if requiresAuth {
            guard let token = await currentAccessToken() else {
                throw APIError.notAuthenticated
            }
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    @MainActor
    private func currentAccessTokenOnMain() -> String? {
        tokenStore.current?.accessToken
    }

    @MainActor
    private func currentRefreshTokenOnMain() -> String? {
        tokenStore.current?.refreshToken
    }

    private func currentAccessToken() async -> String? {
        await currentAccessTokenOnMain()
    }

    private func currentRefreshToken() async -> String? {
        await currentRefreshTokenOnMain()
    }

    private func performWithRefresh(
        request: URLRequest,
        requiresAuth: Bool
    ) async throws -> (Data, URLResponse) {
        let (data, response) = try await sendOnce(request: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIError.transport(message: "non-HTTP response")
        }

        if http.statusCode == 401 && requiresAuth {
            // Try one refresh + retry. A second 401 means the refresh itself
            // either failed or yielded a still-invalid access token; in
            // either case, the user is logged out from this client's view.
            let refreshed = try await runRefresh()
            if !refreshed {
                throw APIError.refreshFailed
            }

            // Rebuild the request with the new access token. We can't just
            // mutate the old `Authorization` header — even though the token
            // string is what changes — because the original request might
            // have come from the caller without an auth header at all.
            var retried = request
            if let token = await currentAccessToken() {
                retried.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }
            return try await sendOnce(request: retried)
        }
        return (data, response)
    }

    private func sendOnce(request: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await session.data(for: request)
        } catch let error as URLError {
            throw APIError.transport(message: error.localizedDescription)
        } catch {
            throw APIError.transport(message: error.localizedDescription)
        }
    }

    // Run the refresh (single-flight). Returns true if refresh succeeded and
    // tokens were rotated; false if there was no refresh token to use.
    // RefreshCoordinator collapses concurrent calls into one shared Task.
    private func runRefresh() async throws -> Bool {
        guard let refreshToken = await currentRefreshToken() else {
            return false
        }
        return try await refresher.run { [self] in
            do {
                let response: AuthResponse = try await self.send(
                    method: "POST",
                    path: "/auth/refresh",
                    body: RefreshBody(refreshToken: refreshToken),
                    requiresAuth: false
                )
                // Hop to MainActor to update TokenStore. We must wait for this
                // to complete before returning so the retried request can read
                // the new access token.
                try await MainActor.run {
                    try self.tokenStore.applyRefresh(
                        accessToken: response.accessToken,
                        refreshToken: response.refreshToken
                    )
                }
                // Persist the rotated tokens to Keychain. Failure here doesn't
                // block the retry — worst case is a relaunch finds the old
                // tokens and triggers another refresh, which is fine.
                try? await self.tokenStore.persistRotated(
                    accessToken: response.accessToken,
                    refreshToken: response.refreshToken
                )
                return true
            } catch {
                // Refresh-token failure (401, network, decode). Clear so UI
                // flips to logged-out and the user re-authenticates.
                await self.tokenStore.clear()
                return false
            }
        }
    }

    private func decode<T: Decodable>(data: Data, response: URLResponse) throws -> T {
        guard let http = response as? HTTPURLResponse else {
            throw APIError.transport(message: "non-HTTP response")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw Self.decodeErrorEnvelope(data: data, status: http.statusCode)
        }
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decoding(error.localizedDescription)
        }
    }

    private static func decodeErrorEnvelope(data: Data, status: Int) -> APIError {
        let decoder = JSONDecoder()
        if let envelope = try? decoder.decode(APIErrorEnvelope.self, from: data) {
            return APIError.server(
                status: status,
                code: envelope.error.code,
                message: envelope.error.message,
                requestId: envelope.error.requestId
            )
        }
        // Backend should always send the envelope, but if a misconfigured
        // proxy returns plain text we still want a structured error.
        let body = String(data: data, encoding: .utf8) ?? "<no body>"
        return APIError.server(status: status, code: "unknown", message: body, requestId: "")
    }
}

// Empty body / response sentinels. Calls that take no body pass `EmptyBody()`;
// calls that return no body use the `sendNoContent` overload. We avoid
// `Void` in the generic position because Void isn't Decodable.
public struct EmptyBody: Codable, Sendable, Equatable {
    public init() {}
}

public struct EmptyResponse: Codable, Sendable, Equatable {
    public init() {}
}

// Actor that ensures exactly one /auth/refresh is in flight at a time.
//
// Why an actor and not a lock? Two reasons:
//   - The work being protected is async (an HTTP call). NSLock or DispatchSemaphore
//     would deadlock if the held thread suspends. Actors are designed for this:
//     they serialize message handling but allow suspension inside the work.
//   - Sendable check: the in-flight Task is itself Sendable, so storing it on
//     an actor and handing references to other tasks is type-safe by construction.
//
// **Why we don't clear `inFlight` in a `defer` after the task finishes**: if a
// leader runs the entire refresh + retry inline before a follower coroutine
// even reaches `run(_:)` — which is what happens under a fast in-process test
// scheduler or any deterministic dispatcher — the follower would see
// `inFlight == nil` and start a SECOND refresh, defeating single-flight.
// The Android port of this code surfaced exactly this bug under JUnit's test
// dispatcher on Linux CI; the iOS code has the same theoretical bug but
// `URLSession.data(for:)` does real I/O so the leader always suspends long
// enough for followers to see the populated `inFlight`. Hardening this here
// before Phase 7 (where the sync engine triggers real 401 bursts under
// arbitrary scheduling) means the property holds under any future runtime
// behavior, not just today's empirical timing.
//
// Replacement strategy: track an explicit `isFinished` alongside the Task.
// Late followers from the same burst find a not-yet-finished Task and join
// it. A brand new 401 (rare — would need a freshly-rotated token to
// immediately yield another 401) sees the finished Task, recognizes it as
// stale, and starts a fresh one. Swift's Task doesn't expose `isFinished`
// directly, so we maintain it ourselves via an Entry struct.
private actor RefreshCoordinator {
    private struct Entry {
        let task: Task<Bool, any Error>
        // Mutable on the actor's isolation domain only — set to true after
        // the leader's `task.value` returns.
        var isFinished: Bool
    }

    private var inFlight: Entry?

    func run(_ work: @Sendable @escaping () async throws -> Bool) async throws -> Bool {
        // Phase 1: decide leader vs follower while holding the actor's
        // isolation. This is a pure synchronous read/write — no `await`,
        // so two concurrent calls can't interleave here.
        if let existing = inFlight, !existing.isFinished {
            // Phase 2 (follower): release isolation by `await`ing the
            // existing Task. Multiple followers on the same Task get the
            // same Bool / same thrown error for free.
            return try await existing.task.value
        }

        // Either no Task ever ran (`inFlight == nil`) or the previous one
        // already finished (stale). Start fresh.
        let task = Task<Bool, any Error> {
            try await work()
        }
        inFlight = Entry(task: task, isFinished: false)

        // Phase 2 (leader): do the actual work. Suspending on `task.value`
        // releases actor isolation; followers can call `run(_:)` and see
        // the populated, not-yet-finished entry. When the work completes
        // (or throws), we re-acquire isolation, mark the entry finished —
        // we deliberately do NOT clear it so a follower that arrived right
        // before `task.value` returned still sees a populated entry to
        // await. The next NEW 401 burst will replace the entry above.
        do {
            let result = try await task.value
            markFinished()
            return result
        } catch {
            markFinished()
            throw error
        }
    }

    private func markFinished() {
        guard var entry = inFlight else { return }
        entry.isFinished = true
        inFlight = entry
    }
}

File: android/app/src/main/java/in/santosh_bharadwaj/sharedlist/core/networking/ApiClient.kt
package `in`.santosh_bharadwaj.sharedlist.core.networking

import `in`.santosh_bharadwaj.sharedlist.core.auth.ApiErrorEnvelope
import `in`.santosh_bharadwaj.sharedlist.core.auth.AuthResponse
import `in`.santosh_bharadwaj.sharedlist.core.auth.RefreshBody
import `in`.santosh_bharadwaj.sharedlist.core.auth.TokenStore
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.HttpClientEngine
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.headers
import io.ktor.client.request.request
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.http.Url
import io.ktor.http.contentType
import io.ktor.http.takeFrom
import io.ktor.serialization.kotlinx.json.json
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json

/**
 * Typed errors surfaced to callers. Mirrors iOS [APIError].
 */
public sealed class ApiError(message: String, cause: Throwable? = null) : Exception(message, cause) {
    public object NotAuthenticated : ApiError("not authenticated") {
        @Suppress("unused")
        private fun readResolve(): Any = NotAuthenticated
    }
    public data class Server(
        val status: Int,
        val code: String,
        val errorMessage: String,
        val requestId: String,
    ) : ApiError("[$status] $errorMessage")
    public data class Decoding(val detail: String) : ApiError("decoding failed: $detail")
    public data class Transport(val detail: String) : ApiError("transport: $detail")
    public object RefreshFailed : ApiError("refresh failed") {
        @Suppress("unused")
        private fun readResolve(): Any = RefreshFailed
    }
}

/**
 * Single point of network access for the Android app. Mirrors iOS APIClient.
 *
 * Responsibilities, in order of importance:
 *   1. Send Serializable bodies, decode Serializable responses, surface typed
 *      errors via [ApiError].
 *   2. Inject `Authorization: Bearer <accessToken>` from [TokenStore].
 *   3. On a 401, refresh the access token AT MOST ONCE PER REQUEST and retry,
 *      with multiple concurrent in-flight requests sharing one refresh.
 *
 * Concept #3 is the **single-flight refresh** pattern — the most important
 * piece of this file. Why it matters:
 *
 *   - Suppose ten ViewModels each fire a request at app foreground.
 *   - The access token has expired since last use.
 *   - All ten get a 401.
 *   - Without single-flight, all ten would call /auth/refresh in parallel.
 *     The backend's refresh endpoint rotates the refresh token (new one
 *     issued, old one marked used). The first request swaps the token, but
 *     the other nine race using the same OLD refresh token. The backend's
 *     reuse detection then revokes ALL of the user's refresh tokens — instant
 *     surprise logout.
 *   - With single-flight: the first 401 starts a single refresh `Deferred`;
 *     every subsequent 401 within the in-flight window awaits the same
 *     `Deferred` and retries with the new access token.
 *
 * Implementation: a `Mutex` plus a single `CompletableDeferred<Boolean>`
 * stored as `inFlight`. The Mutex protects the `inFlight` reference itself,
 * not the work — the actual `/auth/refresh` HTTP call happens OUTSIDE the
 * Mutex so subsequent 401s can read the same `inFlight` and `await()` it.
 *
 * Kotlin notes (vs iOS):
 *   - iOS uses an `actor` to serialize access. Kotlin actors-as-mailboxes are
 *     marked obsolete; the idiomatic primitive is `Mutex.withLock { }`. The
 *     guarantee is the same (one coroutine holds the lock at a time).
 *   - `CompletableDeferred<Boolean>` is the Kotlin analog of Swift's
 *     `Task<Bool, Error>` — both are "the single result of the work" handles
 *     that multiple awaiters can subscribe to.
 *
 * What we deliberately do NOT do:
 *   - Proactively check token expiry before each request. We could decode
 *     the JWT's `exp` claim, but reactive (refresh on 401) handles clock skew,
 *     server clock drift, sleep/wake jitter for free. The cost is one extra
 *     round trip per token rotation — fine.
 */
public class ApiClient(
    private val baseUrl: String,
    private val tokenStore: TokenStore,
    engine: HttpClientEngine = OkHttp.create(),
    /** Long-lived scope for the persisted-rotation work after a refresh. */
    private val ioScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) {

    @PublishedApi
    internal val json: Json = Json {
        ignoreUnknownKeys = true
        // The backend ships ISO-8601 timestamps for fields we'll meet in
        // Phase 7 (`updated_at` etc). Configuring lenient handling now means
        // adding sync DTOs later doesn't require touching the global parser.
        explicitNulls = false
    }

    private val client: HttpClient = HttpClient(engine) {
        install(ContentNegotiation) {
            json(json)
        }
        // No default Authorization header — we inject per-request via
        // `withAuth(...)`, because some calls (signup, login, refresh) MUST
        // NOT carry one. A blanket header would short-circuit refresh logic.
        expectSuccess = false
    }

    /** Mutex that guards [inFlight]; held only while reading/writing the reference. */
    private val refreshLock = Mutex()

    /**
     * The current in-flight refresh, or `null` if none. Multiple coroutines
     * that hit a 401 in the same instant collapse onto this single Deferred.
     */
    private var inFlight: CompletableDeferred<Boolean>? = null

    /**
     * Send a request and decode the response.
     *
     * Generic `T` is the response type (any `@Serializable` data class). The
     * KSerializer is supplied implicitly via reified type parameters — Kotlin
     * doesn't have Swift's free Codable conformance, so callers go through
     * the inline reified wrapper [send] / [sendNoContent] below.
     */
    public suspend inline fun <reified T> send(
        method: HttpMethod,
        path: String,
        body: Any? = null,
        requiresAuth: Boolean = true,
    ): T {
        val response = perform(method, path, body, requiresAuth)
        return try {
            response.body()
        } catch (t: SerializationException) {
            throw ApiError.Decoding(t.message ?: "serialization failed")
        }
    }

    /**
     * 204 / no-content variant — body is read and discarded. Used for the
     * logout endpoint and any future endpoints that don't return a payload.
     */
    public suspend fun sendNoContent(
        method: HttpMethod,
        path: String,
        body: Any? = null,
        requiresAuth: Boolean = true,
    ) {
        perform(method, path, body, requiresAuth)
    }

    /**
     * `@PublishedApi internal` because the inline `send<T>()` above must call
     * it (inline functions can only call same-or-higher-visibility members).
     * Performs a single request with the 401 → refresh → retry policy.
     */
    @PublishedApi
    internal suspend fun perform(
        method: HttpMethod,
        path: String,
        body: Any?,
        requiresAuth: Boolean,
    ): HttpResponse {
        val initial = sendOnce(method, path, body, requiresAuth)
        if (initial.status == HttpStatusCode.Unauthorized && requiresAuth) {
            val refreshed = runRefresh()
            if (!refreshed) throw ApiError.RefreshFailed
            // Retry once with the rotated access token.
            val retried = sendOnce(method, path, body, requiresAuth)
            return validate(retried)
        }
        return validate(initial)
    }

    private suspend fun sendOnce(
        method: HttpMethod,
        path: String,
        body: Any?,
        requiresAuth: Boolean,
    ): HttpResponse {
        return try {
            client.request {
                this.method = method
                url { takeFrom(Url(baseUrl + path)) }
                contentType(ContentType.Application.Json)
                if (requiresAuth) {
                    val token = tokenStore.current?.accessToken ?: throw ApiError.NotAuthenticated
                    headers { append(HttpHeaders.Authorization, "Bearer $token") }
                }
                if (body != null) {
                    setBody(body)
                }
            }
        } catch (e: ApiError) {
            throw e
        } catch (t: Throwable) {
            // Wrap network / DNS / timeout failures in the typed Transport
            // error so callers don't have to import Ktor types to handle them.
            throw ApiError.Transport(t.message ?: t::class.simpleName ?: "transport failure")
        }
    }

    private suspend fun validate(response: HttpResponse): HttpResponse {
        if (response.status.value in SUCCESS_STATUS_RANGE) return response
        // Non-2xx — try to decode the standard envelope. If parsing fails,
        // surface the raw body so a misconfigured proxy doesn't silently turn
        // into "unknown error".
        val raw = try { response.bodyAsText() } catch (_: Throwable) { "" }
        val envelope = try {
            json.decodeFromString<ApiErrorEnvelope>(raw)
        } catch (_: SerializationException) {
            null
        }
        throw if (envelope != null) {
            ApiError.Server(
                status = response.status.value,
                code = envelope.error.code,
                errorMessage = envelope.error.message,
                requestId = envelope.error.requestId,
            )
        } else {
            ApiError.Server(
                status = response.status.value,
                code = "unknown",
                errorMessage = raw.ifBlank { "<no body>" },
                requestId = "",
            )
        }
    }

    /**
     * Coordinated refresh — the heart of the single-flight pattern. Returns
     * `true` if the refresh succeeded and tokens were rotated, `false` if
     * there was no refresh token to use.
     *
     * Two-phase locking:
     *   1. Acquire [refreshLock] briefly to read-or-set [inFlight]. If a
     *      not-yet-completed refresh is in flight, return its Deferred
     *      immediately. If no Deferred exists or the existing one already
     *      completed (stale — it represents a previous burst's result),
     *      replace it with a fresh one and become the leader.
     *   2. Release the lock, do the actual HTTP call OUTSIDE the lock.
     *      Otherwise concurrent 401s would queue on the lock instead of
     *      sharing the same in-flight result.
     *
     * Why we don't null `inFlight` in a `finally` after the work completes:
     * if a leader runs the refresh inline (e.g. under a test scheduler or
     * a fast loopback) before a follower coroutine has even reached this
     * function, the follower would see `inFlight == null` and start a SECOND
     * refresh — defeating the single-flight invariant. By keeping the
     * completed Deferred in place and replacing it only when a new 401 burst
     * arrives, late-arriving followers from the same burst still join. A
     * brand new 401 (after the rotation succeeded but a separate code path
     * still has a stale token, very rare) sees the completed Deferred,
     * recognizes it as stale, and starts a fresh refresh — correct.
     */
    private suspend fun runRefresh(): Boolean {
        val refreshToken = tokenStore.current?.refreshToken ?: return false

        val (deferred, isLeader) = refreshLock.withLock {
            val existing = inFlight
            if (existing != null && !existing.isCompleted) {
                existing to false
            } else {
                val fresh = CompletableDeferred<Boolean>()
                inFlight = fresh
                fresh to true
            }
        }

        if (!isLeader) {
            return deferred.await()
        }

        // We're the leader — actually do the refresh.
        val result = try {
            val response: AuthResponse = send(
                method = HttpMethod.Post,
                path = "/auth/refresh",
                body = RefreshBody(refreshToken = refreshToken),
                requiresAuth = false,
            )
            tokenStore.applyRefresh(response.accessToken, response.refreshToken)
            // Persist outside the critical path — a slow disk write shouldn't
            // delay the retried request.
            ioScope.launch {
                tokenStore.persistRotated(response.accessToken, response.refreshToken)
            }
            true
        } catch (t: Throwable) {
            // Refresh failed (401, transport, decode). Clear so UI flips to
            // logged-out and the user re-authenticates.
            try { tokenStore.clear() } catch (_: Throwable) { /* ignore */ }
            false
        }
        // Complete the Deferred so any follower coroutines awaiting it wake
        // up with the same Boolean. We deliberately do NOT null `inFlight`
        // here — see kdoc rationale above.
        deferred.complete(result)
        return result
    }

    private companion object {
        // Standard "successful response" range. Inlined as a named constant
        // because magic-number literals (200..299) in network code rot fast.
        val SUCCESS_STATUS_RANGE = 200..299
    }
}
